src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java

```diff
@@ -0,0 +1,219 @@
+import org.openkilda.wfm.topology.flowhs.mapper.HaFlowMapper;
+
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import net.jodah.failsafe.RetryPolicy;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class HaFlowReadService {
+    private final HaFlowRepository haFlowRepository;
+    private final FlowRepository flowRepository;
+    private final TransactionManager transactionManager;
+    private final int readOperationRetriesLimit;
+    private final Duration readOperationRetryDelay;
+
+    public HaFlowReadService(@NonNull PersistenceManager persistenceManager,
+                             int readOperationRetriesLimit, @NonNull Duration readOperationRetryDelay) {
+        haFlowRepository = persistenceManager.getRepositoryFactory().createHaFlowRepository();
+        flowRepository = persistenceManager.getRepositoryFactory().createFlowRepository();
+        transactionManager = persistenceManager.getTransactionManager();
+        this.readOperationRetriesLimit = readOperationRetriesLimit;
+        this.readOperationRetryDelay = readOperationRetryDelay;
+    }
+
+    private <T> RetryPolicy<T> getReadOperationRetryPolicy() {
+        return new RetryPolicy<T>()
+                .handle(PersistenceException.class)
+                .withDelay(readOperationRetryDelay)
+                .withMaxRetries(readOperationRetriesLimit)
+                .onRetry(e -> log.debug("Failure in transaction. Retrying #{}...", e.getAttemptCount(),
+                        e.getLastFailure()))
+                .onRetriesExceeded(e -> log.error("Failure in transaction. No more retries", e.getFailure()));
+    }
+
+    /**
+     * Fetches all HA-flows.
+     */
+    public List<HaFlowResponse> getAllHaFlows() {
+        return transactionManager.doInTransaction(getReadOperationRetryPolicy(), haFlowRepository::findAll)
+                .stream()
+                .map(haFlow -> HaFlowMapper.INSTANCE.toHaFlowDto(haFlow, flowRepository, haFlowRepository))
+                .map(HaFlowResponse::new)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Gets HA-flow by ID.
+     */
+    public HaFlowResponse getHaFlow(@NonNull String haFlowId) throws FlowNotFoundException {
+        return transactionManager.doInTransaction(getReadOperationRetryPolicy(), () ->
+                        haFlowRepository.findById(haFlowId))
+                .map(haFlow -> HaFlowMapper.INSTANCE.toHaFlowDto(haFlow, flowRepository, haFlowRepository))
+                .map(HaFlowResponse::new)
+                .orElseThrow(() -> new FlowNotFoundException(haFlowId));
+    }
+
+    /**
+     * Gets HA-flow paths by ID.
+     */
+    public HaFlowPathsResponse getHaFlowPaths(@NonNull String haFlowId) throws FlowNotFoundException {
```

Comment:
This method is too long. Please, consider splitting it into smaller parts with meaningful names just for readability.

Reply:
Completely agree with you. Spent a lot of time thinking about it, but didn't find a reasonable solution. 
Will try to do such refactoring during the diversity part adding.